ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTreeRaoProvider.java

```diff
@@ -67,74 +69,97 @@ public SearchTreeRaoProvider() {
     public CompletableFuture<RaoResult> run(RaoInput raoInput, RaoParameters parameters) {
         RaoUtil.initData(raoInput, parameters);
 
+        // optimization is made on one given state only
         if (raoInput.getOptimizedState() != null) {
-            RaoData raoData = new RaoData(
-                    raoInput.getNetwork(),
-                    raoInput.getCrac(),
-                    raoInput.getOptimizedState(),
-                    raoInput.getPerimeter(),
-                    raoInput.getReferenceProgram(),
-                    raoInput.getGlskProvider(),
-                    raoInput.getBaseCracVariantId(),
-                    parameters.getLoopflowCountries());
-            TreeParameters treeParameters = TreeParameters.buildForPreventivePerimeter(parameters.getExtension(SearchTreeRaoParameters.class), true);
-            return CompletableFuture.completedFuture(new SearchTree().run(raoData, parameters, treeParameters).join());
+            return optimizeOneStateOnly(raoInput, parameters);
         }
 
+        // compute initial sensitivity on all CNECs
+        // this is necessary to have initial flows for MNEC and loopflow constraints on CNECs, in preventive and curative perimeters
+        SystematicSensitivityResult initialSensitivityResult;
+        try {
+            initialSensitivityResult = initialSensitivityAnalysisOnAllPerimeters(raoInput, parameters);
+        } catch (SensitivityAnalysisException e) {
+            return CompletableFuture.completedFuture(new RaoResult(RaoResult.Status.FAILURE));
+        }
+
+        // optimize preventive perimeter
         stateTree = new StateTree(raoInput.getCrac(), raoInput.getNetwork(), raoInput.getCrac().getPreventiveState());
+        Network network = raoInput.getNetwork();
+        network.getVariantManager().cloneVariant(network.getVariantManager().getWorkingVariantId(), PREVENTIVE_STATE);
+        network.getVariantManager().setWorkingVariant(PREVENTIVE_STATE);
+
+        RaoResult preventiveRaoResult = optimizePreventivePerimeter(raoInput, parameters, initialSensitivityResult).join();
 
         if (stateTree.getOptimizedStates().size() == 1) {
-            // only preventive optimization
-            return optimizePreventivePerimeter(raoInput, parameters, null, true);
-        } else {
-            // Compute initial sensitivity on all CNECs (preventive and curative)
-            // This is necessary to have initial flows for MNEC and loopflow constraints on CNECs in the curative perimeter
-            // No need to recompute the sensitivities in the beginning of the preventive RAO
-            new InitialSensitivityAnalysis(
-                    raoInput.getNetwork(),
-                    raoInput.getCrac(),
-                    raoInput.getReferenceProgram(),
-                    raoInput.getGlskProvider(),
-                    parameters).run();
+            return CompletableFuture.completedFuture(preventiveRaoResult);
```

Comment:
this is not the same behavior as before since the result will not be joignable (in line 92 you use .join(), so line 95 won't be executed until the RAO is complete)

Reply:
Indeed, I am changing this